qa/tasks/rsync.py

```diff
@@ -0,0 +1,270 @@
+
+        self.fs = Filesystem(self.ctx)
+
+        self.snap_enable = bool(self.config.get('snapenable', False))
+
+        #Get CephFS mount client and mount object
+        if len(self.ctx.mounts.items()):
+            flag = False
+            for i, j in sorted(self.ctx.mounts.items()):
+                tmp = int(i)
+                if tmp == self.mount_point:
+                    self.mount_point = tmp
+                    self.my_mnt = j
+                if (not self.mount_point) and (not flag):
+                    self.mount_point = tmp
+                    self.my_mnt = j
+                    flag = True
+        else:
+            assert len(self.ctx.mounts.items()) > 0, 'No mount available asserting rsync'
+
+        #Set source directory for rsync
+        if self.config.get('data_dir'):
+            self.work_unit = True
+            self.source_dir = misc.get_testdir(self.ctx) + '/mnt.{}'.format(self.mount_point) + \
+                              '/{}/'.format(self.config.get('data_dir'))
+        else:
+            self.data_dir = misc.get_testdir(self.ctx) + '/mnt.{}/'.format(self.mount_point) + 'source'
+            self.source_dir = self.data_dir + '/subdir'
+
+    def _run(self):
+        try:
+            self.do_rsync()
+        except:
+            # Log exceptions here so we get the full backtrace (it's lost
+            # by the time someone does a .get() on this greenlet)
+            self.logger.exception("Exception in do_rsync:")
+            raise
+
+    def stop(self):
+        self.stopping.set()
+
+    #Function to check directory exists.
+    def check_if_dir_exists(self, path):
+        """
+            Call this to check stat of directory.
+            :return: True if directory exists.
+                     False if not.
+        """
+        try:
+            self.my_mnt.stat(path)
+            return True
+        except Exception, e :
+            logging.error(e)
+            return False
+
+    def do_rsync(self):
+
+        iteration = 0
+        finished = False
+
+        # Create destination directory
+        self.my_mnt.run_shell(["mkdir", "rsyncdir"])
+
+        if self.snap_enable:
+            # Enable snapshots
+            self.fs.mon_manager.raw_cluster_cmd("mds", "set", "allow_new_snaps", "true", "--yes-i-really-mean-it")
+
+        if not self.work_unit:
+            # Create a data directory, sub directory and rsync directory
+            self.my_mnt.run_shell(["mkdir", "{}".format(self.data_dir)])
+            self.my_mnt.run_shell(["mkdir", "{}".format(self.source_dir)])
+            should_stop = False
+
+        #Check for source directory exists
+        while not self.check_if_dir_exists(self.source_dir):
+            time.sleep(5) # if source dorectory not exists wait for 5s and poll
+            iteration += 1
+            if iteration > 5:
+                assert self.check_if_dir_exists(self.source_dir), 'assert, source Directory doesnot exists'
```

Comment:
1. This will always assert, because `self.check_if_dir_exists(self.source_dir)` is already `false` when `iteration > 5`.
2. Double calling of `self.check_if_dir_exists(self.source_dir)` in the same iteration.
3. I think only the `assert` statement is required here without the while loop, by asserting before the directory creation at L170. And the function `check_if_dir_exists` could be avoided (exception and returning). 
4. Not sure what is achieved by waiting for 5s here. Is there any other thread creates the source directory in the mean time?
4. typo: s/dorectory/directory

Reply:
The mkdir for the source directory may not have completed / i.e directory not yet created, hence the sleep of 5 secs. Will fix the typo